cs3api4lab/cs3_file_api.py

```diff
@@ -0,0 +1,237 @@
+
+        self.client_id = config['client_id']
+        self.client_secret = config['client_secret']
+
+        # prepare the gRPC connection
+
+        if secure_channel:
+
+            try:
+
+                cert = open(config['client_cert'], 'rb').read()
+                key = open(config['client_key'], 'rb').read()
+
+                ca_cert = open(config['ca_cert'], 'rb').read()
+                credentials = grpc.ssl_channel_credentials(ca_cert, key, cert)
+
+                ch = grpc.secure_channel(reva_host, credentials)
+
+            except():
+                print("Error create secure channel")
+                raise
+
+        else:
+            ch = grpc.insecure_channel(reva_host)
+
+        self.cs3_stub = cs3gw_grpc.GatewayAPIStub(ch)
+        return
+
+    def _authenticate(self, userid):
+
+        # ToDo: use real authentication data or get token from author provider
+        # authReq = cs3gw.AuthenticateRequest(type='bearer', client_secret=userid)
+
+        if userid not in self.tokens or self.tokens[userid]['exp'] < time.time():
+            auth_req = cs3gw.AuthenticateRequest(type='basic', client_id=self.client_id, client_secret=self.client_secret)
+            auth_res = self.cs3_stub.Authenticate(auth_req)
+            self.tokens[userid] = {'tok': auth_res.token, 'exp': time.time() + self.authtokenvalidity}
+
+        # piggy back on the opportunity to expire old tokens, but asynchronously
+        # as to not impact the current session: let's use python3.7's coroutines support
+        # asyncio.run(_async_cleanuptokens())
+
+        return self.tokens[userid]['tok']
+
+    def _cs3_reference(self, endpoint, fileid):
+
+        if fileid[0] == '/':
+            # assume this is a filepath
+            return cs3spr.Reference(path=fileid)
+
+        if endpoint == 'default' or endpoint is None:
+            raise IOError('A CS3API-compatible storage endpoint must be identified by a storage UUID')
+
+        # assume we have an opaque fileid
+        return cs3spr.Reference(id=cs3spr.ResourceId(storage_id=endpoint, opaque_id=fileid))
+
+    def stat(self, endpoint, fileid, userid):
+
+        """
+        Stat a file and returns (size, mtime) as well as other extended info using the given userid as access token.
+        Note that endpoint here means the storage id. Note that fileid can be either a path (which MUST begin with /), or an id (which MUST NOT
+        start with a /).
+        """
+
+        time_start = time.time()
+
+        ref = self._cs3_reference(endpoint, fileid)
+
+        stat_info = self.cs3_stub.Stat(request=cs3sp.StatRequest(ref=ref), metadata=[('x-access-token', self._authenticate(userid))])
+
+        time_end = time.time()
+        self.log.info('msg="Invoked stat" fileid="%s" elapsedTimems="%.1f"' % (fileid, (time_end - time_start) * 1000))
+
+        if stat_info.status.code == cs3code.CODE_OK:
+            self.log.debug('msg="Stat result" data="%s"' % stat_info)
+            return {
```

Comment:
As I said in a previous review, I'm pretty sure Jupyter also need creation time and wether it's RO or RW.
The permissions are especially important and you have `permission_set` that should give you everything you need.
The creation time, maybe doesn't break if we re-use the mtime. I will create an issue in the CS3APIs to add this.

Reply:
Notebook using last modified timestamp in checkpoint, sorting files, browser cache time, check if user try to save old version of file. Created time in not important parts (I'm don't see in notebook code). But when Riva will support create time stamp we will using this attribute.